### PR TITLE
fix password protect downloads and copying ComicInfo files in LocalSource

### LINKED ISSUE #1076: Can't download when `Password protect downloads` is on
### Steps to reproduce

1. Set CBZ archive password
2. Enable `Password protect downloads`
3. Try downloading chapter

### Expected behavior

Downloads normally with password applied to the cbz archive.

### Actual behavior

Gives `java.io.FileNotFoundException`. But only when `Password protect downloads` is on. When turned off, it downloads normally but the cbz is of course not encrypted.

### Crash logs

[logs.txt](https://github.com/jobobby04/TachiyomiSY/files/14276390/logs.txt)

### Tachiyomi version

1.10.2

### Android version

Android 9

### Device

Samsung Note 8

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with an extension, I should be opening an issue in the extension's repository.
- [X] I have gone through the [FAQ](https://tachiyomi.org/docs/faq/general) and [troubleshooting guide](https://tachiyomi.org/docs/guides/troubleshooting/).
- [X] I have updated the app to version **[1.10.2](https://github.com/jobobby04/tachiyomisy/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by Shamicen**
fixes #1076

Trying to access the image files directly causes access denied errors, switching to streams fixes this.

### CODE REVIEW

**Comment by Shamicen** [comment]
I also fixed copying ComicInfo files into encrypted archives. This is still broken for unencrypted ComicInfo files but I will check if this is also broken upstream and fix it in Mihon first if it is.

**Comment by Shamicen** [comment]
It has already been fixed upstream. I applied the fix here because it was only a few lines of code.

File: source-local/src/androidMain/kotlin/tachiyomi/source/local/LocalSource.kt [hunk 1]
```diff
@@ -312,14 +311,26 @@ actual class LocalSource(
...
             zipEncrypted.addStream(comicInfoFileStream, zipParameters)
-            return zipEncrypted.file
+
+            val realZipEncrypted = folder.createFile(COMIC_INFO_ARCHIVE)!!
+            realZipEncrypted.openOutputStream().use { out ->
+                zipEncryptedFile.inputStream().use {
+                    it.copyTo(out)
+                }
+            }
+            zipEncryptedFile.delete()
```
Comment (FooIbar) [inline source-local/src/androidMain/kotlin/tachiyomi/source/local/LocalSource.kt:327]: Ditto.

File: app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt [hunk 2]
```diff
@@ -673,10 +673,14 @@ class Downloader(
...
+            zipParameters.fileNameInZip = img.name
+            zip.addStream(
+                img.openInputStream(),
+                zipParameters,
+            )
+        }
+
         zip.close()
 
         val realZip = mangaDir.createFile("$dirname.cbz$TMP_DIR_SUFFIX")!!
```
[the remark this answers is not in the document]
Comment (Shamicen) [reply app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt:686]: > May use the `SeekableByteChannel` overload to write to `realZip` directly

Zip4j the library we use for encrypted zip files does not support creating files from `SeekableByteChannel`.

File: source-local/src/androidMain/kotlin/tachiyomi/source/local/LocalSource.kt [hunk 3]
```diff
@@ -239,14 +239,13 @@ actual class LocalSource(
...
+                    val copiedFile = mangaDir?.let { copyComicInfoFileFromArchive(chapterArchives, it) }
 
-                    val copiedFile = copyComicInfoFileFromArchive(chapterArchives, folderPath)
                     // SY -->
                     if (copiedFile != null && copiedFile.name != COMIC_INFO_ARCHIVE) {
-                        setMangaDetailsFromComicInfoFile(copiedFile.inputStream(), manga)
+                        setMangaDetailsFromComicInfoFile(copiedFile.openInputStream(), manga)
                     } else if (copiedFile != null && copiedFile.name == COMIC_INFO_ARCHIVE) {
-                        val comicInfoArchive = ZipFile(copiedFile)
+                        val comicInfoArchive = ZipFile(copiedFile.filePath)
```
[the remark this answers is not in the document]
Comment (Shamicen) [reply source-local/src/androidMain/kotlin/tachiyomi/source/local/LocalSource.kt]: I used the tempFileManager here.

**Comment by Shamicen** [comment]
I think we should keep the tempFileManager for TachiyomiSY, since zip4j does not support channels. We could still use channels for normal unencrypted zip files if we only switch to using zip4j if opening the file throws an error.

File: app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (FooIbar) [reply app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt:686]: `ZipOutputStream` works too.

**Comment by Shamicen** [comment]
I implemented it using streams but reading them is pretty slow. This is probably worth the load time in `LocalSource` were the cacheDir could get spammed with temp files. But this will not work for The `ZipPageLoader`, my [test implementation](https://github.com/Shamicenso/TachiyomiSY/tree/test_zip_page_loader_streams) had a loading time of 10 seconds per page. If anyone has any ideas how to load them faster please let me know.

**Comment by Shamicen** [comment]
I did some more tests and i don't think using `ZipInputStreams` will be feasible for the `ZipPageLoader`. Every time a page is loaded all pages that came before must be read again (and i think decrypted). that alone adds an overhead of 0.5-1s per page that has come before. I'll implement a 2 way solution were we use the `TempFileManger` for encrypted zip files and channels for unencrypted files. This should be pretty easy to implement since we are already using a 2 way solution with zip4j for encrypted and Tachiyomis Java zip implementation for unencrypted files. 

I think this PR can be merged for now.

**Pull request merged**